### ISSUE TITLE
SolutionTransfer: Implement more general interpolation algorithms

### ISSUE THREAD

**Issue by kronbichler** [dealii/dealii#15225]
Right now, the interpolation done by `SolutionTransfer` entirely works in reference coordinates, https://github.com/dealii/dealii/blob/be2729222ec604ef4403ee641a1b81d3f75576eb/source/numerics/solution_transfer.cc#L514-L526 and similar for the distributed solution transfer https://github.com/dealii/dealii/blob/be2729222ec604ef4403ee641a1b81d3f75576eb/source/distributed/solution_transfer.cc#L459-L468 However, manifolds can change the local metrics and thus the reference-cell interpolation can be lower accuracy. This phenomenon has been observed in a slightly different context in #7357. Inspired by the non-nested multigrid algorithms that do a consistent search operation (see #15148), we should extend the infrastructure in `SolutionTransfer` as well. At least for elements supported by `FEPointEvaluation`, we have reasonably fast methods to perform the interpolation. And once we are able to do interpolation, we should also be able to perform projection operations upon refinement, i.e., compute `(v_new, u_new)_new = (v_new, o_old)_new` written as bilinear form, where `u_old` on the right hand side is evaluated at quadrature points that are the intersection between the refined and un-refined parts of the mesh. The latter would be well-supported by the matrix-free framework.

As I envision that we might have different application requirements, we should make the `SolutionTransfer` classes flexible in the sense that a user can control more closely what algorithm to run, by giving some suitable interface. Opinions?

**Comment by marcfehling** [comment]
> As I envision that we might have different application requirements, we should make the `SolutionTransfer` classes flexible in the sense that a user can control more closely what algorithm to run, by giving some suitable interface. Opinions?

I agree in making the `SolutionTransfer` class more flexible. Though in a different context, @Rombur and I were discussing approaches starting with #11132, where we contemplated on how to do a transfer from a `FE_Nothing` to a different element that does not represent nullspace. I hope we can work on the prototype in the upcoming hackathon.

I would suggest to let `SolutionTransfer` only provide the external scaffolding for the whole transfer operation. For the rework, I suggest we realise the preparation and recovery of the data purely via callback functions: The user needs to specify how the data will be prepared/recovered by providing a function for each operation themselves. `SolutionTransfer` then calls these callbacks at the required place. Ideally, these callbacks will be interchangeable between serial and parallel implementations of `SolutionTransfer`.

We can provide a library of such functions for different purposes: one for preparation and one for recovery of cell-related data (as now with `CellDataTransfer`), dof-related data (as now with `SolutionTransfer`), quadrature-related data (as now with `ContinuousQuadratureDataTransfer`), and with some finite element activation logic (as now with `FieldTransfer`).

We now have four different classes in the parallel::distributed namespace that kind of do the same thing, and I would like to help to merge them into a more flexible class. This will make the code more easily maintainable, more flexible, and avoids code duplication. This would also lay out the path for your vision, @kronbichler.

The question of course remains if the idea I have in my head will actually work out...
